drivers/counter/counter_ace_v1x_art.c

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <device.h>
+#include <devicetree.h>
+#include <zephyr/drivers/counter.h>
+#include <soc.h>
+#include <ace_v1x-regs.h>
+#include <counter/counter_ace_v1x_art_regs.h>
+
+#if defined(__XCC__)
+#include <xtensa/config/core.h>
+#if XCHAL_HAVE_HIFI4 && (defined(__ALDERLAKE__) \
+	|| defined(__METEORLAKE__) || defined(__METEORLAKE_S__))
+#include <xtensa/tie/xt_hifi4.h>
+#elif XCHAL_HAVE_HIFI3
+#include <xtensa/tie/xt_hifi3.h>
+#endif
+#endif
+
+static struct k_spinlock lock;
+
+#define ACE_TSCTRL	(0x72040)
+
+static void counter_ace_v1x_art_ionte_set(bool new_timestamp_enable)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
+	val.u.part.ionte                = new_timestamp_enable;
+	((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full = val.u.full;
+}
+
+static void counter_ace_v1x_art_cdmas_set(uint32_t cdmas)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
+	val.u.part.cdmas                = cdmas;
+	((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full = val.u.full;
+}
+
+static void counter_ace_v1x_art_ntk_set(bool new_timestamp_taken)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
+	val.u.part.ntk                = new_timestamp_taken;
+	((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full = val.u.full;
+}
+
+static uint32_t counter_ace_v1x_art_ntk_get(void)
+{
+	return ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.part.ntk;
+}
+
+static void counter_ace_v1x_art_hhtse_set(bool enable)
+{
+	struct DfTSCTRL val;
+
+	val.u.full                      = ((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full;
+	val.u.part.hhtse                = enable;
+	((volatile struct DfTSCTRL*)(ACE_TSCTRL))->u.full = val.u.full;
+}
+
+static uint64_t counter_ace_v1x_art_counter_get(void)
+{
+#if defined(__XCC__) /* use Xtensa coprocessor for atomic read of 64 bit value */
+	return *((volatile ae_int64 * const)ACE_ARTCS);
```

Comment:
This deeply, deeply creeps me out.  First, just as syntax: that's not what this code means.  I buy that xcc might emit a HiFi instruction to do the load here, but all the C standard requires is that you copy 64 bits of memory via some means or another.  If you want HiFi, please emit a HiFi load explicitly.

And even then: is a HiFi load atomic?  The PIF bus is still 32 bit on these CPUs.  How do you know the HiFi unit isn't just doing two sequential transfers?  Even if they're done as a single transaction on the bus, how do you know the **counter** hardware (itself probably on a 32 bit bus) is interlocking the counter state between the two reads?  Is any of this documented anywhere?

Just... yikes.  If you can get the hardware folks to show up here and guarantee that this usage is correct and atomic all the way down into the counter state, then just tell me I'm wrong.  Otherwise, can we please implement this in a way that can be verified as pure software?  I'm fine with the HiFi load, but just do it twice in a loop to validate that the high word didn't change.

Reply:
Unfortunately, I cannot make any guarantees. You might be right. There is no need to use HiFi at all, if we still need validation. This was just a hint I got and it's purpose was solely to avoid reading the high value twice. Removed.